tests/bluetooth/bsim_bt/bsim_test_adv_resume/src/bs_bt_utils.c

```diff
@@ -0,0 +1,179 @@
+	printk("connected\n");
+}
+
+void wait_disconnected(void)
+{
+	SET_FLAG(flag_is_connected);
+	WAIT_FOR_FLAG_UNSET(flag_is_connected);
+	printk("disconnected\n");
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	UNSET_FLAG(flag_is_connected);
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err != 0) {
+		return;
+	}
+
+	SET_FLAG(flag_is_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+void bs_bt_utils_setup(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+	ASSERT(!err, "bt_enable failed.\n");
+}
+
+static bt_addr_le_t last_scanned_addr;
+
+static void scan_connect_to_first_result__device_found(const bt_addr_le_t *addr, int8_t rssi,
+						       uint8_t type, struct net_buf_simple *ad)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	/* We're only interested in connectable events */
+	if (type != BT_HCI_ADV_IND && type != BT_HCI_ADV_DIRECT_IND) {
+		FAIL("Unexpected advertisement type.");
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+	printk("Got scan result, connecting.. dst %s, RSSI %d\n", addr_str, rssi);
+
+	err = bt_le_scan_stop();
+	ASSERT(!err, "Err bt_le_scan_stop %d", err);
+
+	struct bt_conn *conn;
+
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	ASSERT(!err, "Err bt_conn_le_create %d", err);
+
+	/* Save address for later comparison */
+	memcpy(&last_scanned_addr, addr, sizeof(last_scanned_addr));
+}
+
+void scan_connect_to_first_result(void)
+{
+	int err;
+
+	printk("start scanner\n");
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, scan_connect_to_first_result__device_found);
+	ASSERT(!err, "Err bt_le_scan_start %d", err);
+}
+
+static void scan_expect_same_address__device_found(const bt_addr_le_t *addr, int8_t rssi,
+						   uint8_t type, struct net_buf_simple *ad)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+	char expected_addr_str[BT_ADDR_LE_STR_LEN];
+
+	/* We're only interested in connectable events */
+	if (type != BT_HCI_ADV_IND && type != BT_HCI_ADV_DIRECT_IND) {
+		FAIL("Unexpected advertisement type.");
+	}
+
+	if (!bt_addr_le_eq(&last_scanned_addr, addr)) {
+		bt_addr_le_to_str(&last_scanned_addr, expected_addr_str, sizeof(expected_addr_str));
+		bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+		FAIL("Expected advertiser with addr %s, got %s\n", expected_addr_str, addr_str);
+	}
+	PASS("Advertiser used correct address on resume\n");
+}
+
+void scan_expect_same_address(void)
+{
+	int err;
+
+	printk("start scanner\n");
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, scan_expect_same_address__device_found);
+	ASSERT(!err, "Err bt_le_scan_start %d", err);
+}
+
+static void disconnect_device(struct bt_conn *conn, void *data)
+{
+	int err;
+	struct bt_conn_info info;
+
+	bt_conn_get_info(conn, &info);
+
+	uint8_t role = *(uint8_t *)data;
+
+	if (role != info.role) {
+		return;
+	}
+
+	SET_FLAG(flag_is_connected);
```

Comment:
Why are we setting this flag here?

Reply:
because I cheat and don't have a flag per-connection. Since this fn waits until it is unset (in the disconnected callback), i have to set it again for the next iteration.

Reply:
I see. Please comment such non-obvious behavior :)